torch/csrc/tensor/python_tensor.cpp

```diff
@@ -0,0 +1,274 @@
+  ((PyObject*)&type)->ob_refcnt = 1;
+  ((PyObject*)&type)->ob_type = &metaclass;
+  type.tp_basicsize = sizeof(PyTensorType);
+  type.tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE;
+  type.tp_name = name;
+  type.tp_new = Tensor_new;
+  if (PyType_Ready(&type) < 0) {
+    throw python_error();
+  }
+  if (PyDict_Merge(type.tp_dict, tp_dict, 0) < 0) {
+    throw python_error();
+  }
+}
+
+static const char* get_module(Backend backend) {
+  switch (backend) {
+    case kCPU: return "torch";
+    case kCUDA: return "torch.cuda";
+    case kSparseCPU: return "torch.sparse";
+    case kSparseCUDA: return "torch.cuda.sparse";
+    default: runtime_error("invalid backend: %s", toString(backend));
+  }
+}
+
+static std::string get_name(Backend backend, ScalarType scalarType) {
+  std::ostringstream ss;
+  ss << get_module(backend) << "." << at::toString(scalarType) << "Tensor";
+  return ss.str();
+}
+
+static void set_type(PyTensorType& type_obj, Backend backend, ScalarType scalarType) {
+  auto baseType = globalContext().type_registry[static_cast<int>(backend)][static_cast<int>(scalarType)].get();
+  type_obj.aten_type = baseType ? torch::autograd::VariableType::getType(*baseType) : nullptr;
+  type_obj.is_cuda = backend == kCUDA || backend == kSparseCUDA;
+  type_obj.is_sparse = backend == kSparseCPU || backend == kSparseCUDA;
+}
+
+static void set_name(PyTensorType& type_obj, const std::string& name) {
+  size_t n = sizeof(type_obj.name);
+  strncpy(type_obj.name, name.c_str(), n);
+  type_obj.name[n - 1] = '\0';
+}
+
+static THPObjectPtr get_variable_dict() {
+  auto autograd = THPObjectPtr(PyImport_ImportModule("torch.autograd"));
+  if (!autograd) throw python_error();
+
+  auto variable_class = THPObjectPtr(PyObject_GetAttrString(autograd.get(), "Variable"));
+  if (!variable_class) throw python_error();
+
+  auto variable_type = (PyTypeObject*)variable_class.get();
+  TORCH_ASSERTM(variable_type->tp_base, "missing base type for Variable");
+
+  auto res = THPObjectPtr(PyDict_New());
+  if (!res) throw python_error();
+
+  if (PyDict_Merge(res.get(), variable_type->tp_dict, 0) < 0) {
+    throw python_error();
+  }
+  if (PyDict_Merge(res.get(), variable_type->tp_base->tp_dict, 0) < 0) {
+    throw python_error();
+  }
+
+  return res;
+}
+
+static std::vector<PyTensorType> tensor_types;
+
+static void initialize_aten_types(std::vector<PyTensorType>& tensor_types) {
+  // includes CUDA types even when PyTorch is not built with CUDA
+  auto declared_types = torch::utils::all_declared_types();
+  tensor_types.resize(declared_types.size() + 1);
+
+  for (size_t i = 0, end = declared_types.size(); i != end; i++) {
+    auto& tensor_type = tensor_types[i];
+    Backend backend = declared_types[i].first;
+    ScalarType scalar_type = declared_types[i].second;
+    set_type(tensor_type, backend, scalar_type);
+    set_name(tensor_type, get_name(backend, scalar_type));
+  }
+
+  set_type(tensor_types.back(), kCPU, kFloat);
```

Comment:
it's difficult to follow how the original default tensor type is actually initialized (i.e. this just matches up with what we know it is, but doesn't actually set it).  Can we move the initialization in here?